k9mail/src/main/java/com/fsck/k9/controller/ImapSyncInteractor.java

```diff
@@ -0,0 +1,260 @@
+        this.account = account;
+        this.folderName = folderName;
+        this.listener = listener;
+        this.controller = controller;
+    }
+
+    void performSync(MessageDownloader messageDownloader) {
+        Exception commandException = null;
+
+        try {
+            Timber.d("SYNC: About to process pending commands for account %s", account.getDescription());
+
+            try {
+                controller.processPendingCommandsSynchronous(account);
+            } catch (Exception e) {
+                controller.addErrorMessage(account, null, e);
+                Timber.e(e, "Failure processing command, but allow message sync attempt");
+                commandException = e;
+            }
+
+            Timber.v("SYNC: About to get local folder %s and open it", folderName);
+            localFolder = SyncUtils.getOpenedLocalFolder(account, folderName);
+            localFolder.updateLastUid();
+
+            Store remoteStore = account.getRemoteStore();
+            Timber.v("SYNC: About to get remote folder %s", folderName);
+            Folder remoteFolder = remoteStore.getFolder(folderName);
+            if (!(remoteFolder instanceof ImapFolder)) {
+                throw new IllegalArgumentException("A non-IMAP account was provided to ImapSyncInteractor");
+            }
+            imapFolder = (ImapFolder) remoteFolder;
+
+            if (!SyncUtils.verifyOrCreateRemoteSpecialFolder(account, folderName, imapFolder, listener, controller)) {
+                return;
+            }
+
+            QresyncParamResponse qresyncParamResponse;
+            Timber.v("SYNC: About to open remote IMAP folder %s", folderName);
+            qresyncParamResponse = imapFolder.open(Folder.OPEN_MODE_RW, localFolder.getUidValidity(),
+                    localFolder.getHighestModSeq());
+
+            boolean qresyncEnabled = qresyncParamResponse != null;
+            List<String> expungedUids = new ArrayList<>();
+            if (Expunge.EXPUNGE_ON_POLL == account.getExpungePolicy()) {
+                Timber.d("SYNC: Expunging folder %s:%s", account.getDescription(), folderName);
+                if (qresyncEnabled) {
+                    expungedUids = imapFolder.expungeUsingQresync();
+                } else {
+                    imapFolder.expunge();
+                }
+            }
+
+            int remoteMessageCount = imapFolder.getMessageCount();
+            if (remoteMessageCount < 0) {
+                throw new IllegalStateException("Message count " + remoteMessageCount + " for folder " + folderName);
+            } else {
+                Timber.v("SYNC: Remote message count for folder %s is %d", folderName, remoteMessageCount);
+            }
+
+            handleUidValidity();
+            int newMessages;
+            if (!qresyncEnabled) {
+                NonQresyncSyncInteractor syncInteractor = new NonQresyncSyncInteractor(account, localFolder, imapFolder,
```

Comment:
Changed to `QresyncExtensionHandler`, `NonQresyncExtensionHandler` and `continueSync`.